hibernate-core/src/main/java/org/hibernate/loader/hql/QueryLoader.java

```diff
@@ -450,9 +449,24 @@ protected Object getResultColumnOrRow(
 		if ( hasScalars ) {
 			String[][] scalarColumns = scalarColumnNames;
 			int queryCols = queryReturnTypes.length;
+			int managedTypeIndex = 0;
+			SessionFactoryImplementor factory = session.getFactory();
 			resultRow = new Object[queryCols];
 			for ( int i = 0; i < queryCols; i++ ) {
-				resultRow[i] = queryReturnTypes[i].nullSafeGet( rs, scalarColumns[i], session, null );
+				// Types for inverse relationships have a zero column span in which case we have to get the
+				// result from the row array.
+				Type queryReturnType = queryReturnTypes[i];
+				if ( queryReturnType.getColumnSpan( factory ) == 0 ) {
+					resultRow[i] = row[managedTypeIndex];
+				}
+				else {
+					resultRow[i] = queryReturnType.nullSafeGet( rs, scalarColumns[i], session, null );
+				}
+				// The row array contains entity objects, we have to count up the index to access inverse relationship
+				// entities from the row array
+				if ( queryReturnType.isEntityType() ) {
```

Comment:
What about other managed types? I.e. selecting an embeddable in a  scalar query? Is that even possible?